fix: storage class name conflict

DamiaSan opened the pull request:
#### Problem:
We use `longhorn-<vmimage name>` as storage class name. The StorageClass doesn't have namesapce, but vmimage has. If there are same vmimage name in different namespace, storage class name will be conflicts.

#### Solution:
Use `namespace` and image `UID` as name format.

#### Related Issue(s):
https://github.com/harvester/harvester/issues/5165

- - -

# Test plan:

## Basic Tests

### Case 1: Reproducer from #5165
1. Create two VirtualMachineImages: `default/test-image` and `harvester-public/test-image`
2. After the two VirtualMachineImages are ready, delete `default/test-image`
3. Create a VirtualMachine from VirtualMachineImage `harvester-public/test-image`

## Regression Tests

### Case 1: VM basic operations (create / backup /restore) without error.
1. Create a VMImage.
2. Create a VM without error.
3.. Write some data.
4. Setup backup-target.
5. Backup the VM.
6. Restore a new VM. New VM can start without error.

### Case 2: VMImage with a same name in different namespaces can work
1. Create a VMImage `thisistest` in `default` and `harvester-public` namespaces.
2. Both images can be created without error.

### Case 3: An old VMImage can work without error.
1. Create a Harvester cluster v1.4.0.
2. Create a VMImage.
3. Replace harvester and harvester-webhook deployment with this PR.
4. The VMImage can work without error.

### Case 4: Create VM Template without error
1. Follow case 1.
2. Generate Template with data.
3. After template is ready, launch a new VM and we can see default image name starts with `templateversion-`. New VM can start without error and data is in the new VM.

## Rancher Tests

### Case 1: Guest Cluster with Reproducer from #5165
1. Install Rancher, register Harvester cluster in Rancher
2. Create two VirtualMachineImages: `default/test-image` and `harvester-public/test-image`
3. After the two VirtualMachineImages are ready, delete `default/test-image`
4. Create a new guest cluster on Harvester from VirtualMachineImage `harvester-public/test-image`

Commit:
fix: storage class name conflict

Now the storage class name is created with VMI's UUID to avoid
conflicts that happen when using vmimages with the same name but
with different namespace.

Related to https://github.com/harvester/harvester/issues/5165

Signed-off-by: Damiano Cipriani <damiano.cipriani@suse.com>

w13915984028 commented on pkg/image/backingimage/backingimage.go:
```diff
@@ -174,7 +179,12 @@ func (bib *Backend) createBackingImage(vmi *harvesterv1.VirtualMachineImage) err
+		return err
+	}
+
+	if cachedSC, _ := bib.scCache.Get(storageClassName); cachedSC != nil && cachedSC.DeletionTimestamp != nil {
```
should we return earlier if `cachedSc != nil && cachedSC.DeletionTimestamp == nil` ?

w13915984028 replied:
The reconciller might run to this function might times.

Either use the existing one, or compare parameters and if not same, update it or delete/create new one and log related message/event. WDYT?

DamiaSan commented on pkg/image/backingimage/backingimage.go:
```diff
@@ -83,7 +83,12 @@ func (bib *Backend) deleteBackingImage(vmi *harvesterv1.VirtualMachineImage) err
+	storageClassName, err := util.GetImageStorageClassName(bib.scCache, vmi)
+	if err != nil {
+		return err
+	}
```
The `IsNotFound` error is handled inside the function `GetImageStorageClassName`, do you suggest something different?

w13915984028 replied:
skip the `isNotFound` returned from `bib.scClient.Delete(storageClassName, &metav1.DeleteOptions{}` ?

Vicente-Cheng commented:
Hi @FrankYang0529, 
Please also check this PR. I will take a look at it, too.

w13915984028 commented on pkg/util/image.go:
```diff
@@ -72,11 +72,47 @@ func GetBackingImageDataSourceName(backingImageCache ctllhv1.BackingImageCache,
+	// If not, try with a new name format based on the image UID, which can avoid name conflict. Because
+	// the legacy name is based on image name, so two images with the same name but in different namespaces
+	// would generate the same storage class name (StorageClass doesn't have namespace).
+	sc, err := storageClassCache.Get(legacyStorageClassName(image.Name))
```
a legacy VMI named `abc` in namespace `ns1` has SC name  `longhorn-abc`, 
a new VMI `abc` in namespace `ns2` will have SC name `longhorn-af3069-....` in new mode

the `GetStorageClass` tries legacy first, `ns2/abc` will get `longhorn-abc`

by contrast, try new mode first,  `ns1/abc` won't get `longhorn-af3069-...`

DamiaSan replied:
> ns2/abc will get longhorn-abc

Is this a problem? Because we should simply avoid conflicts ...

DamiaSan replied:
What do you think @FrankYang0529 @Vicente-Cheng ?

FrankYang0529 replied:
I agree that we should get new mode first. A potential issue is like following:

* User creates a new VMImage `default/test`.
* User creates a new VMIMage `new-ns/test`. In current logic, it will remove `longhorn-test` StorageClass and cause issue.

Vicente-Cheng commented:
> Can anyone enlighten me why we're creating a separate storage class for each VMImage? Why can't we just re-use the default storage class unless the user specifies otherwise?

This is the special behavior of Longhorn backingimage. We need a specific field, like `parameters/backingImage`, to point out how we create the PVC from the backing image. 

That's why we need to do that.

FrankYang0529 commented on pkg/api/vm/handler.go:
```diff
@@ -717,32 +717,46 @@ func getMigrationUID(vmi *kubevirtv1.VirtualMachineInstance) string {
+				}
+			}
+		}
+	}()
```
We need to remove `vmt` as well if it is not nil.

WebberHuang1118 commented on pkg/api/vm/handler.go:
```diff
@@ -776,7 +790,27 @@ func (h *vmActionHandler) createTemplate(namespace, name string, input CreateTem
+			}
+			return h.sanitizeVolumes(vmtv, pvcMap, vmImageMap)
+		})
+		if err != nil {
```
This function has become too long; I suggest factoring this section out into a separate function.

m-ildefons replied:
I've refactored and simplified this functions quite a bit. I've also moved it and all related functions into their own file, which should make them much easier to review.
I haven't yet gone through all of the related functions to refactor them as well, but as far as I can tell most of them aren't quite as convoluted.

WebberHuang1118 reviewed (changes requested):
In `pkg/api/vm/handler.go`, the change introduces nested `if` and `for` statements, which reduces readability and may make future maintenance or extensions more difficult. 

Please factor out the logic and wrap it into smaller, more focused functions.

m-ildefons committed:
refactor: break up createTemplate

Break up `createTemplate` function from the VM action handler into
several smaller functions and move it and its associated functions into
a separate file.

The `createTemplate` function now only handles a few very basic tasks,
which are all implemented in their own separate functions. This makes
each function much easier to understand and it simplifies the code-paths
in the error cases.

In case the `createTemplate` function encounters an error in any of its
actions, it will clean up both the VirtualMachineTemplateVersion as well
as the VirtualMachineTemplate that it created before.

Moving this group of functions into a separate file makes it easier to
navigate the source code.

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

m-ildefons committed:
break up getPVCStorageClassMap function

Break up overly complicated implementation of the
`getPVCStorageClassMap` function into smaller easier to understand
parts.

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

m-ildefons committed:
break up createVMImages function

Factor out the creation of each individual VMImage.
Move decision logic for the backingImage to its own function.

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

w13915984028 reviewed (commented):
Could we add some unit test code to cover the target cases as more as possible? thanks.

m-ildefons commented:
> Oh, I thought you were addressing the comment from @w13915984028. I’ll check your update — thanks.

Not yet, I will do that. But I have addressed your findings.

> This can also happen when directly applying manifests, e.g., using kubectl.

Yes, I noticed. That's why I'm pushing this PR forward and not trying to fix the Terraform provider only.

WebberHuang1118 commented:
Re: [discussion](https://github.com/harvester/harvester/pull/9392#discussion_r2692911126)

To support VM root disks with third-party storage, we established `VMImage` as an [abstraction layer](https://github.com/harvester/harvester/pull/7170) in `v1.5.0` to accommodate different backends.

As a result, instead of directly accessing `VMImage` member fields, e.g.,

```
return GenerateStorageClassName(string(image.UID)), nil
```

```
sc, err := storageClassCache.Get(legacyStorageClassName(image.Name))
```

We should access it via the `VMIOperator`, which is intended to provide a clean and encapsulated interface. I realize some legacy code does not follow this principle, but we should gradually remove such implementations. Since this is a new enhancement, IMO, we must adhere to the interface design.

I envision extending `VMIOperator` with a new method, `GetSCName()`. This would allow any part of the source code holding a `VMIOperator` stub to retrieve the StorageClass name.

WebberHuang1118 commented on pkg/api/vm/template.go:
```diff
@@ -0,0 +1,449 @@
+
+func (h *vmActionHandler) createSecrets(templateVersion *harvesterv1.VirtualMachineTemplateVersion, vm *kubevirtv1.VirtualMachine) error {
+	for index, credential := range vm.Spec.Template.Spec.AccessCredentials {
+		if sshPublicKey := credential.SSHPublicKey; sshPublicKey != nil && sshPublicKey.Source.Secret != nil {
```
This `if` expression is too long; please wrap it in a separate function.

WebberHuang1118 commented on pkg/api/vm/template.go:
```diff
@@ -0,0 +1,449 @@
+				return err
+			}
+		}
+		if userPassword := credential.UserPassword; userPassword != nil && userPassword.Source.Secret != nil {
```
ditto

WebberHuang1118 commented on pkg/api/vm/template.go:
```diff
@@ -0,0 +1,449 @@
+// If the PVC has an annoation, which associates it with a BackingImage, then
+// this function will return the StorageClass name of the BackingImage. Otherwise it
+// will return the StorageClass name of the PVC itself.
+func (h *vmActionHandler) getStorageClassNameFromPVCOrBackingImage(pvc *corev1.PersistentVolumeClaim) (scName string, err error) {
```
The function name is too long. Suggest shortening it.

WebberHuang1118 commented on pkg/api/vm/template.go:
```diff
@@ -0,0 +1,449 @@
+
+// Given an ImageID, find the corresponding VMImage and figure out it's
+// StorageClass name
+func (h *vmActionHandler) getStorageClassNameFromImageID(imageId string) (scName string, err error) {
```
The function name is too long. Suggest shortening it.

WebberHuang1118 commented on pkg/api/vm/template.go:
```diff
@@ -0,0 +1,449 @@
+	vmImage := vmImageMap[vmImageName]
+
+	claim := pvcMap[claimName]
+	targetStorageClassName := util.GenerateStorageClassName(string(vmImage.UID))
```
The variable name is too long. Suggest shortening it.

m-ildefons committed:
VM Image Operator: Storage Class Utility

Implement storage class utilities in the VM Image Operator.
The VM Image Operator is an interface providing a collection of
utilities for working with VirtualMachineImages.
The utilities for figuring out the storage class name (both in the
legacy format and the improved format) are implemented in this interface
instead of their own utility functions.

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

m-ildefons committed:
Image Operator: Shorten Storage Class Name

Shorten storage class name for VM Images such that the new UID-based
format stays under the 40 character limit of names for entities of
Longhorn.

This prevents log-spam on every operation, including lookups of storage
classes.

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

m-ildefons committed:
VM Images: fix review comments

Reorganize code in the VM Image API handler to satisfy review comments.

Re-order the usage of the storage class naming scheme in the VM Image
Operator's GetStorageClassName method.
The legacy name is now returned if and only if no storage class with a
name fitting the new naming scheme exists and if there exists a storage
class matching the old naming scheme.

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

WebberHuang1118 commented on pkg/util/storageclass.go:
```diff
@@ -0,0 +1,18 @@
+		return longhorn.DataEngineType(v)
+	}
+	return longhorn.DataEngineTypeAll
+}
```
thanks

WebberHuang1118 commented on pkg/api/vm/template.go:
```diff
@@ -0,0 +1,472 @@
+	x_4a322c04 := h.vmio.GetStorageClassName(&vmImage)
+	if _, ok := pvc.Annotations[cdicommon.AnnCreatedForDataVolume]; ok {
+		x_4a322c04 = *claim.Spec.StorageClassName
+	}
```
ditto

Vicente-Cheng replied:
ditto, maybe `targetSCName`?

olblak replied:
Please see my comment [above](https://github.com/harvester/harvester/pull/9392#discussion_r2798045181)

WebberHuang1118 reviewed (commented):
Thanks for the enhancement. The implementation looks clear and compact from my perspective—please address or reply to the comments.

I think it’s close to merge. As discussed, please make sure all test cases are covered before merging. I’ll also have testing when I am available.

WebberHuang1118 commented on pkg/api/vm/template.go:
```diff
@@ -0,0 +1,472 @@
+// Get Storage Class Name from ImageID
+// Given an ImageID, find the corresponding VMImage and figure out it's
+// StorageClass name
+func (h *vmActionHandler) x_bcdb3011(imageId string) (scName string, err error) {
```
IMO, we still need a meaningful function name. For example, `getSCNameFromImage` could be a good reference.

olblak replied:
Could we stop arguing on function name and stay focus to land this pull request?
Naming is hard, and we all agree that `x_bcdb3011` from @m-ildefons is a bad one.
To be honest, I don't understand why `targetStorageClassName` is a bad one neither.
I don't know what too long, really means.

@WebberHuang1118 Feel free to make a suggestion for a better name if you have a strong opinion on the matter, but to me, it seems nit picking.

This kind of [comment](https://github.com/harvester/harvester/pull/9392#discussion_r2791315245) are more constructive  IMHO

WebberHuang1118 replied:
> To be honest, I don't understand why targetStorageClassName is a bad one neither.
> I don't know what too long, really means.
> This kind of https://github.com/harvester/harvester/pull/9392#discussion_r2791315245 are more constructive IMHO

IMO, if it’s an internal variable in the code (e.g., a local variable or an internal struct field), using `targetSCName` is totally reasonable.

However, as I mentioned in [this comment](https://github.com/harvester/harvester/pull/9392#discussion_r2792574338), that’s just my suggestion and it’s open for discussion.

During the earlier conversation, I thought we were still discussing it, so out of respect I didn’t want to come across as telling the author what to do. That said, if there’s any follow-up feedback or further discussion, I’d be happy to share my thoughts.

I don’t have a strong opinion on it—my only point is that the naming should remain meaningful.

m-ildefons commented on pkg/api/vm/template.go:
```diff
@@ -0,0 +1,472 @@
+}
+
+// Copy Credentials Secret
+func (h *vmActionHandler) x_a0fc0b55(templateVersion *harvesterv1.VirtualMachineTemplateVersion, index int, credential *kubevirtv1.AccessCredential) (err error) {
```
Any suggestions? Maybe `copyCredentialsSecret`?

WebberHuang1118 replied:
LGTM for `copyCredentialsSecret`.

olblak replied:
Please see my comment [above](https://github.com/harvester/harvester/pull/9392#discussion_r2798045181)

m-ildefons committed:
backing-image-controller test: fixup

Fix up VM Image Operator instantiation in the Backing Image controller
tests.
Add the missing fake client and fake cache to be able to call the VM
Image Operator constructor.

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

m-ildefons commented on pkg/api/vm/template.go:
```diff
@@ -0,0 +1,474 @@
+}
+
+func (h *vmActionHandler) createVMTemplate(namespace string, input CreateTemplateInput) (vmt *harvesterv1.VirtualMachineTemplate, err error) {
+	vmt, err = h.vmTemplateClient.Create(
```
If a VM Template with the same name in the same namespace already exists, then the `h.vmTemplateClient.Create(...)` function call at line 397 will return an error. `createTemplate` already handles this case by aborting and propagating the error upwards.
We could of course check whether the error from `h.vmTemplateClient.Create(...)` is a HTTP 409 Conflict error. In that case we could possibly return the existing VM Template without creating a new one. Then the handler can continue with creating a new VM Template Version for that existing VM Template. Would that work?

IMO, in general, there is too much logic in the VM Template handler. A lot of this logic should probably exist in the VM Template controller instead.
However, I think re-designing the VM Template handler and controller is out of scope for this PR.

m-ildefons commented on pkg/api/vm/template.go:
```diff
@@ -0,0 +1,474 @@
+	name := fmt.Sprintf("%s-%s", vmt.Name, rand.String(6))
+	vmtID := fmt.Sprintf("%s/%s", vmt.Namespace, vmt.Name)
+	vmSourceSpec := h.sanitizeVirtualMachineForTemplateVersion(name, vm)
+	vmtv, err = h.vmTemplateVersionClient.Create(
```
This is a bit tricky, a conflict is unlikely, because of the randomly generated name. But I'll add a check an let the handler retry with a different randomly generated name in case the name is already occupied.

m-ildefons replied:
`cleanupVMTemplate()` is called _after_ `cleanupVMTemplateVersion()` in case something went wrong during the template creation.
Should `cleanupVMTemplateVersion()` fail, then we will indeed leak a resource and require manual intervention. However that case is very unlikely and the only scenario I can think of is if the Kubernetes API server becomes unavailable between the API calls. However in that case we can't do any cleanup anyways.

m-ildefons committed:
fixup image controller tests

Fix up image controller tests.

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

m-ildefons committed:
API Handler: Remove unused function

Remove unused function updateVMVolumeClaimTemplate from VM API handler

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

w13915984028 reviewed (commented):
some questions,

and the new file `template.go` looks to refactor some existing code logic, anyway to add some test codes to cover the different cases as more as possible? thanks.

@m-ildefons Do you have a detailed local test report?

The main concern is: the refactored code touchs many normal and abnormal running paths, we need enough tests to ensure they are well covered.

m-ildefons committed:
VM Handler: Add cleanup and error handling

Add cleanup functions and error handling cases to VM action handler.

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

m-ildefons committed:
fixups after rebase

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

m-ildefons committed:
utils: owner reference generator

Add a general purpose generator function for owner references to the
utils package.
This function generates the owner reference from any object struct so
the object can become owner of another object.
The function propagates errors in case a struct is not a suitable object
struct and otherwise returns an owner reference struct.

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

m-ildefons committed:
VM action handler: Tests and refactor

Make use of the general purpose owner reference generator.

Move sanitizeVirtualMachineForTemplateVersion to template.go, since it
is only used there.

Reorganize and simplify createTemplateWithData. It's unnecessary to
track storage class names in a map, because they can be easily
marshalled from PVCs and PVCs are already tracked in a map.

Simplify createVMImages and createVMImage such that fewer parameters
need to be passed between functions.

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

m-ildefons commented on pkg/api/vm/template.go:
```diff
@@ -0,0 +1,474 @@
+		return nil
+	}
+
+	if err = h.vmTemplateClient.Delete(vmt.Namespace, vmt.Name, &metav1.DeleteOptions{}); err != nil {
```
I'm not sure I fully understand what you want here.
If I do an explicit lookup with a `h.vmTemplateClient.Get(...)` then I still have a brief period of time between the return of that function and the `h.vmTemplateClient.Delete` call where the object might disappear from the API server.

But if you meant that I should check if `err` contains an `IsNotFound` error after the call to `h.vmTemplateClient.Delete`, that is checked just below this line.

Besides, the VM Template object passed as argument to this function is only non-nil if it has been created before. Not sure why you think it might not exist?

m-ildefons committed:
VM template handler: fixes

- Fix Image backend logic. Default to BackingImage in case provisioner
  is Longhorn, but no data engine is explicitly specified

- Fix map initialization. Initialize Golang maps before writing to them.

- Make consistent use of GetOwnerReferenceFor() from utils.

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

m-ildefons committed:
VM action handler: apply gofmt -w -s

Apply gofmt simplification rules to VM action handler tests.

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

m-ildefons committed:
Unit Tests: Add tests for VM template action

Add two unit tests for important functions used in the VM template
action API handler: GetVMImageBackend and GetStorageClassName.

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

w13915984028 commented on pkg/api/vm/template.go:
```diff
@@ -0,0 +1,580 @@
+	// target storage class name
+	targetSCName := h.vmio.GetStorageClassName(&vmImage)
+	if _, ok := pvc.Annotations[cdicommon.AnnCreatedForDataVolume]; ok {
+		targetSCName = *claim.Spec.StorageClassName
```
similar: is `claim.Spec.StorageClassName != nil` checked before?

w13915984028 reviewed (commented):
Some last questions/suggestions, thanks.

m-ildefons committed:
VM template API handler: error paths

Fix some minor details on the error paths of the VM template API
handler:
- In a deferred function, log an error only in case `err != nil`
- Don't return a partially constructed PVC map in case of an error
- Check the `.Spec.StorageClassName` attribute of a PVC for nil
- `make()` a map for annoations

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

m-ildefons committed:
VM API handler: handle default storage class

Handle default storage class when generating PVC templates while
creating a VM template

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>
